L1Trigger/TrackFindingTMTT/interface/GlobalCacheTMTT.h

```diff
@@ -0,0 +1,56 @@
+#ifndef L1Trigger_TrackFindingTMTT_GlobalCacheTMTT_h
+#define L1Trigger_TrackFindingTMTT_GlobalCacheTMTT_h
+
+// Data shared across all streams by TMTT L1 tracking.
+//
+// Provides the python configuration parameters
+// & optional histogramming/debugging.
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "L1Trigger/TrackFindingTMTT/interface/Settings.h"
+#include "L1Trigger/TrackFindingTMTT/interface/HTrphi.h"
+#include "L1Trigger/TrackFindingTMTT/interface/StubWindowSuggest.h"
+#include "L1Trigger/TrackFindingTMTT/interface/TrackerModule.h"
+#include "L1Trigger/TrackFindingTMTT/interface/Histos.h"
+
+#include <list>
+#include <memory>
+
+namespace tmtt {
+
+  class GlobalCacheTMTT {
+  public:
+    GlobalCacheTMTT(const edm::ParameterSet& iConfig)
+        : settings_(iConfig),              // Python configuration params
+          htRphiErrMon_({0., 0, 0, 0}),    // rphi HT error monitoring
+          stubWindowSuggest_(&settings_),  // Recommend FE stub window sizes.
+          hists_(&settings_)               // Histograms
+    {
+      hists_.book();
+    }
+
+    // Get functions
+    Settings& settings() const { return settings_; }
+    HTrphi::ErrorMonitor& htRphiErrMon() const { return htRphiErrMon_; }
+    StubWindowSuggest& stubWindowSuggest() const { return stubWindowSuggest_; }
+    const std::list<TrackerModule>& listTrackerModule() const { return listTrackerModule_; }
+    Histos& hists() const { return hists_; }
+
+    // Set functions
+    void setListTrackerModule(const std::list<TrackerModule>& list) const {
```

Comment:
This method can not be `const`, because it is not thread safe and it modifies the the object. Especially how it use used in https://github.com/cms-sw/cmssw/blob/de9b0a8d8582aba8fa9c9f221c0e233879735d8e/L1Trigger/TrackFindingTMTT/plugins/TMTrackProducer.cc#L136
is not thread safe.

Reply:
The function setListTrackerModule() stores its argument inside GlobalCacheTMTT the first time it is called. So it will store this info from whichever thread calls it first. (We don't care which thread wins, since they all call it with the same argument). Would it suffice if I add a mutex lock inside function setListTrackerModule(), to eliminate the risk that two threads call it at the same time?

Reply:
A mutex (or an atomic) would make it safe-enough for one Run per job (i.e. non-rundependent MC). If this will be part of the planned redesign, I would be fine with the mutex for the purposes of 11_1.

I would encourage to design the next version to avoid mutexes.

Reply:
mutex lock added to setListTrackerModule().